### PR TITLE
denite.nvim -> ddu.vim

### PR DESCRIPTION
**Pull request by creasty**
Crucial for migration:

- [x] https://github.com/Shougo/ddu-ui-ff/issues/17
- [x] https://github.com/Shougo/ddu-ui-ff/issues/18

Nice to have:

- https://github.com/Shougo/ddu-ui-ff/pull/19
- https://github.com/Shougo/ddu-ui-ff/pull/20
- https://github.com/Shougo/ddu-ui-ff/issues/22

Minor issues:

- https://github.com/Shougo/ddu-ui-ff/issues/21
- https://github.com/Shougo/ddu-ui-ff/issues/23
- C-r to preserve the input while refreshing items

### CODE REVIEW

**Comment by 4513ECHO** [comment]
Hello, your dotfiles looks nice 😄 
Your original sources have already exist: 

- [rg source](https://github.com/shun/ddu-source-rg)
- [external command (fd)](https://github.com/matsui54/ddu-source-file_external)
- [ghq source](https://github.com/4513ECHO/ddu-source-ghq)

How about use these?

**Comment by creasty** [comment]
@4513ECHO Cheers!

> How about use these?

Yes, I'm aware of those plugins. However, they didn't fully satisfy my needs so I had to copy and revise.

Take ghq for instance, I wanted to display the path that's relative to the ghq's root path, which isn't supported by the existing plugin.
https://github.com/creasty/dotfiles/blob/c4e17ce84a859c8c4ca2c2820a22171e91e36614/nvim/denops/%40ddu-sources/ghq.ts#L64-L66

For rg, I added an ability to read input directly from the prompt, and also improved word highlight by going through all sub-matches.
https://github.com/creasty/dotfiles/blob/c4e17ce84a859c8c4ca2c2820a22171e91e36614/nvim/denops/%40ddu-sources/rg.ts#L119

https://github.com/creasty/dotfiles/blob/c4e17ce84a859c8c4ca2c2820a22171e91e36614/nvim/denops/%40ddu-sources/rg.ts#L66

Regarding fd, perhaps I could use the external command plugin.

But overall, I think ddu's architecture is quite straightforward to build upon and feel like it's much effortless to create extensions that are locally optimized for one person. ;)

**Comment by 4513ECHO** [comment]
> But overall, I think ddu's architecture is quite straightforward to build upon and feel like it's much effortless to create extensions that are locally optimized for one person. ;)

Yes, it is true. ddu is realy great plugin!

I am  maintainer of `ddu-source-ghq`, so I will try adding `"relative"` param which make path the relative of ghq's root.

Thank you for reaction of my suggestion!

**Comment by creasty** [comment]
> I am maintainer of `ddu-source-ghq`, so I will try adding `"relative"` param which make path the relative of ghq's root.

Oh, didn't notice! 😆 
Glad that I could help, haha

File: nvim/dein.toml [hunk 1]
```diff
@@ -136,6 +139,39 @@ let g:NERDTreeAutoDeleteBuffer = 1
 let g:loaded_nerdtree_fs_menu = 1
 '''
 
+#  ddu
+#-----------------------------------------------
+[[plugins]] # Dark deno-powered UI framework for neovim/Vim8
```
Comment (creasty) [inline nvim/dein.toml:144]: No lazy initialization was necessary.

```sh-session
$ nvim --headless --startuptime /dev/stdout -c quit | grep 'deno\|ddu'
015.154  000.078  000.078: sourcing /Users/creasty/.config/nvim/autoload/user/plugin/ddu.vim
015.969  000.092  000.092: sourcing /Users/creasty/.config/nvim/dein/.cache/init.vim/.dein/autoload/ddu/custom.vim
016.600  000.116  000.116: sourcing /Users/creasty/.config/nvim/dein/.cache/init.vim/.dein/autoload/ddu.vim
053.394  000.072  000.072: sourcing /Users/creasty/.config/nvim/dein/.cache/init.vim/.dein/plugin/denops/debug.vim
053.736  000.081  000.081: sourcing /Users/creasty/.config/nvim/dein/.cache/init.vim/.dein/autoload/denops.vim
053.839  000.363  000.283: sourcing /Users/creasty/.config/nvim/dein/.cache/init.vim/.dein/plugin/denops.vim
```

```sh-session
$ repeat 5 ( time nvim --headless -c quit ; sleep 0.1 )
nvim --headless -c quit  0.11s user 0.08s system 147% cpu 0.127 total
nvim --headless -c quit  0.11s user 0.08s system 145% cpu 0.130 total
nvim --headless -c quit  0.12s user 0.08s system 146% cpu 0.134 total
nvim --headless -c quit  0.11s user 0.07s system 144% cpu 0.130 total
nvim --headless -c quit  0.11s user 0.08s system 145% cpu 0.130 total
```

**Pull request merged**